### ISSUE TITLE
Implement Quality Focal Loss (QFL) in YOLOX

### ISSUE THREAD

**Issue by iumyx2612** [open-mmlab/mmdetection#8016]
In the original implementation of QFL, it's calculate per scale level
```
        losses_cls, losses_bbox, losses_dfl,\
            avg_factor = multi_apply(
                self.loss_single,
                anchor_list,
                cls_scores,
                bbox_preds,
                labels_list,
                label_weights_list,
                bbox_targets_list,
                self.prior_generator.strides,
                num_total_samples=num_total_samples)
```
But in YOLOX, the loss are calculate as a whole without using multi_apply
```
        (pos_masks, cls_targets, obj_targets, bbox_targets, l1_targets,
         num_fg_imgs) = multi_apply(
             self._get_target_single, flatten_cls_preds.detach(),
             flatten_objectness.detach(),
             flatten_priors.unsqueeze(0).repeat(num_imgs, 1, 1),
             flatten_bboxes.detach(), gt_bboxes, gt_labels)

        # The experimental results show that ‘reduce_mean’ can improve
        # performance on the COCO dataset.
        num_pos = torch.tensor(
            sum(num_fg_imgs),
            dtype=torch.float,
            device=flatten_cls_preds.device)
        num_total_samples = max(reduce_mean(num_pos), 1.0)

        pos_masks = torch.cat(pos_masks, 0)
        cls_targets = torch.cat(cls_targets, 0)
        obj_targets = torch.cat(obj_targets, 0)
        bbox_targets = torch.cat(bbox_targets, 0)
        if self.use_l1:
            l1_targets = torch.cat(l1_targets, 0)

        loss_bbox = self.loss_bbox(
            flatten_bboxes.view(-1, 4)[pos_masks],
            bbox_targets) / num_total_samples
        loss_obj = self.loss_obj(flatten_objectness.view(-1, 1),
                                 obj_targets) / num_total_samples
        loss_cls = self.loss_cls(
            flatten_cls_preds.view(-1, self.num_classes)[pos_masks],
            cls_targets) / num_total_samples
```
Can I calculate QFL target, then pass all of the concat prediction and target to QFL without using multi_apply
```
        loss_cls = self.loss_cls(
            flatten_cls_preds.view(-1, self.num_classes), (joint_labels, joint_scores),
            weight=1.0,
            avg_factor=num_total_cls_samples)
```

**Comment by RangiLyu** [comment]
> Can I calculate QFL target, then pass all of the concat prediction and target to QFL without using multi_apply

Of course, you can.
